### PR TITLE
set/get affinity

### PR DESCRIPTION
**Pull request by kuebk**
Lets you get/set affinity on the given pid.

Mentioned back in time here: https://github.com/joyent/libuv/issues/160

### CODE REVIEW

File: include/uv.h [hunk 1]
```diff
@@ -946,7 +946,14 @@ typedef struct uv_process_options_s {
...
   char* cwd;
-
+  /*
+   * Array of cores to run the process on
+   */
+  char* affinity;
+  /*
+   * Sizeof affinity
+   */
+  int affinity_size;
```
Comment (ry) [inline include/uv.h:956]: Needs docs on how these should be set for default scheduling

File: include/uv.h [hunk 2]
```diff
@@ -982,6 +989,15 @@ UV_EXTERN int uv_spawn(uv_loop_t*, uv_process_t*,
...
 
+/*
+ * Sets the cpu affinity on the given process pid
+ */
+UV_EXTERN int uv_set_affinity(int pid, char *core, int core_size);
+
+/*
+ * Gets the cpu affinity of the given process pid
+ */
+UV_EXTERN int uv_get_affinity(int pid, char **core, int *core_size);
```
Comment (ry) [inline include/uv.h:1000]: Do we really need to add to `process_options_t` if we have these?

Comment (ry) [inline include/uv.h:1000]: - should be with the type. `char**`. What should the `core` parameter be set to? It's unclear how this works. More docs?

Comment (kuebk) [inline include/uv.h:1000]: If you want `uv_spawn` to do all the work then yes we need to modify `process_options_t`.

**Comment by kuebk** [comment]
Test included.

**Comment by ry** [comment]
@bnoordhuis @piscisaureus can you guys review this?

**Comment by piscisaureus** [comment]
Should it not return uv_err_t so it can properly report errors?

**Comment by bnoordhuis** [comment]
I suppose my first question is: why this feature? Why not let the OS figure it out?

**Comment by kuebk** [comment]
We've experienced better results when we have set affinity on our own.

**Comment by bnoordhuis** [comment]
Can you quantify that? Another thing: this patch is linux specific.

**Comment by pietern** [comment]
For IPC on a NUMA system, I've seen ~2x performance when the processes were set to run on neighboring cores.

**Comment by bnoordhuis** [comment]
After mulling it over, I've decided we won't take this patch.

@pietern: I suspect that you're looking at a badly tuned system or a bug in the OS scheduler if you see performance differences like that. (The stock linux 3.x scheduler for example has exactly that bug in power-saving mode.)

**Pull request closed without merging**